contracts/contracts/DC.sol

```diff
@@ -0,0 +1,376 @@
+        _register(name, msg.sender, secret, ensPrice);
+        // Update Name Record and send events
+        uint256 tokenId = uint256(keccak256(bytes(name)));
+        NameRecord storage nameRecord = nameRecords[bytes32(tokenId)];
+        nameRecord.renter = msg.sender;
+        nameRecord.lastPrice = price;
+        nameRecord.timeUpdated = uint32(block.timestamp);
+        if (bytes(url).length > 0) {
+            nameRecord.url = url;
+        }
+
+        lastRented = name;
+
+        emit NameRented(name, msg.sender, price, url);
+
+        // Return any excess funds
+        uint256 excess = msg.value - (baseRentalPrice + ensPrice);
+        if (excess > 0) {
+            (bool success, ) = msg.sender.call{value: excess}("");
+            require(success, "cannot refund excess");
+        }
+    }
+
+    /**
+     * @dev `_register` calls RegistrarController register and is used to register a name
+     * it is passed a value to cover the costs of the ens registration
+     * @param name The name to be registered e.g. for test.country it would be test
+     * @param owner The owner address of the name to be registered
+     * @param secret A secret passed by the client e.g. const [secret] = useState(Math.random().toString(26).slice(2))
+     * @param ensPrice the price paid to ens for the registration
+     */
+    function _register(
+        string calldata name,
+        address owner,
+        bytes32 secret,
+        uint256 ensPrice
+    ) internal whenNotPaused {
+        bytes[] memory emptyData;
+        registrarControllerContract.register{value: ensPrice}(
+            name,
+            owner,
+            duration,
+            secret,
+            resolver,
+            emptyData,
+            reverseRecord,
+            fuses,
+            wrapperExpiry
+        );
+    }
+
+    /**
+     * @dev `renew` calls RegistrarController renew and is used to renew a name
+     * this also takes a fee for the web2 renewal which is held by DC.sol a check is made to ensure the value sent is sufficient for both fees
+     * duration is set at the contract level
+     * @param name The name to be registered e.g. for test.country it would be test
+     * @param url A URL that can be embedded in a web2 default domain page e.g. a twitter post
+     */
+    function renew(
+        string calldata name,
+        string calldata url,
+        address owner
+    ) public payable whenNotPaused {
+        require(bytes(name).length <= 128, "DC: name too long");
+        require(bytes(url).length <= 1024, "DC: url too long");
+        uint256 ensPrice = getENSPrice(name);
+        uint256 price = baseRentalPrice + ensPrice;
+        require((price) <= msg.value, "DC: insufficient payment");
+        _renew(name, ensPrice);
+
+        // Update Name Record and send events
+        uint256 tokenId = uint256(keccak256(bytes(name)));
+        NameRecord storage nameRecord = nameRecords[bytes32(tokenId)];
```

Comment:
Need to check nameRecord is owned by msg.sender. Can't rely on ENS's function (_renew) since someone could have registered the domain in ENS but not here